src/hooks/SocialAuth.ts

```diff
@@ -0,0 +1,104 @@
+import { randomBytes } from 'crypto'
+import useLocalStorage from 'hooks/LocalStorage'
+import usePopupWindow from 'hooks/PopupWindow'
+import { useEffect, useState } from 'react'
+
+type OAuthData = {
+  code: string
+}
+
+type OauthOptions = {
+  client_id: string
+  scope: string
+  response_type?: 'code' | 'token'
+  code_challenge?: 'challenge'
+  code_challenge_method?: 'plain'
+}
+
+const useOauthPopupWindow = (url: string, oauthOptions: OauthOptions) => {
+  const [csrfToken, setCsrfToken] = useLocalStorage(
+    `oauth_csrf_token_${oauthOptions.client_id}`,
+    randomBytes(16).toString('hex'),
+    true,
+  )
+
+  const redirectUri = process.env.NEXT_PUBLIC_TWITTER_REDIRECT_URI as string
+
+  oauthOptions.response_type = oauthOptions.response_type ?? 'code'
+
+  const state = `${oauthOptions.client_id};${csrfToken}`
+
+  // prettier-ignore
+  const { onOpen, windowInstance } = usePopupWindow(
+    `${url}?${Object.entries(oauthOptions).map(([key, value]) => `${key}=${value}`).join("&")}&redirect_uri=${encodeURIComponent(redirectUri)}&state=${encodeURIComponent(state)}`
+  )
+  const [error, setError] = useState(null)
+  const [authData, setAuthData] = useState<OAuthData | null>(null)
+  const [isAuthenticating, setIsAuthenticating] = useState<boolean>(false)
+
+  /** On a window creation, we set a new listener */
+  useEffect(() => {
+    if (!windowInstance) return
+
+    const windowInstanceOpenInitially = !windowInstance.closed
+
+    window.localStorage.removeItem('oauth_popup_data')
+    setIsAuthenticating(true)
+
+    new Promise<OAuthData>((resolve, reject) => {
+      const interval = setInterval(() => {
+        try {
+          const { data, type } = JSON.parse(
+            window.localStorage.getItem('oauth_popup_data') as string,
```

Comment:
**potential issue** using localstorage as a messaging system might not be the best approach. I think the standard practice here is to use the native `postMessage` api: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage.

This would save us from having to do all this polling, and potential security/reliability pitfalls from relying on localstorage.

**suggestion**

1. add an event listener on this parent window object

   ```js
   window.addEventListener('message', () => )
   ```

1. In the popup window (so, in `auth/success/index.page.tsx`), we can add JS to send a message to the parent window to tell it to close.

```js
const data = '' // TODO whatever oauth data (`code` etc.) we need to pass back to the app.
window.opener.postMessage(data);

window.close() // tell the window to close itself. No need for polling.
```

1. Finally, back in our original `message` event handler, we do the final stuff (like persisting verification state to firebase.

Happy to sketch this out in further detail if you need.

P.s. if it's helpful, here's an example of what I'm suggesting:

- `message` event listener: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/81126/diffs#da076f04ccf03ffb0511cf2b9badf2f85db3c74e_0_57
- JS that executes in the auth/success/index.page.tsx. This JS sends `message` back to parent window with all the oauth data, and closes itself: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/81126/diffs#62fb64bed6fc2c5e4f0fb412061b551b1de22b06_0_10

Reply:
This is great, definitely down to use this instead. Again, a lot of the frontend OAuth code was taken from guild.xyz so if there's a better approach to anything here i'm open to it.